File: solutions/go/luhn/1/luhn.go
// Package luhn provides Luhn checksums checks.
package luhn

import (
	"strconv"
	"strings"
	"unicode"
)

// Valid checks whether a given number is valid according to Luhn.
func Valid(s string) bool {
	numberString := strings.Replace(s, " ", "", -1)
	numbers := []rune(numberString)

	if len(numbers) < 2 {
		return false
	}

	valid := make([]int, len(numbers))
	for i, r := range numbers {
		if !(unicode.IsDigit(r) || r == ' ') {
			return false
		}
		valid[i], _ = strconv.Atoi(string(r))
	}

	for i := len(valid) - 2; i >= 0; i -= 2 {
		doubled := 2 * valid[i]
		if doubled > 9 {
			doubled -= 9
		}
		valid[i] = doubled
	}

	sum := 0
	for _, i := range valid {
		sum += i
	}

	check := sum % 10
	if check == 0 {
		return true
	}

	return false
}

File: solutions/julia/luhn/1/luhn.jl
function isvalid(c::Char)
     48 <= Int(c) < 58 || false
end

function over(x::Int)
    doubled = 2 * x
    if doubled > 9
        doubled -= 9
    end
    doubled
end

"""
Predicate to check whether string is a valid Luhn number
"""
function luhn(s::String)
    numberstring = replace(s, r"\s+" => s"")
    length(numberstring) < 2 && return false
    all(isvalid(c) for c in numberstring) || return false
    valid = [parse(Int, c) for c in numberstring]
    for i in length(valid)-1:(-2):1
        valid[i] = over(valid[i])
    end
    s = sum(valid)
    s % 10 == 0 || false
end
